specs/phase1/validator.md

```diff
@@ -0,0 +1,462 @@
+
+##### Custody key reveals
+
+Up to `MAX_CUSTODY_KEY_REVEALS`, [`CustodyKeyReveal`](./custody-game.md#custodykeyreveal) objects can be included in the `block`. The custody key reveals must satisfy the verification conditions found in [custody key reveal processing](./custody-game.md#custody-key-reveals). The validator receives a small reward for each custody key reveal included.
+
+##### Early derived secret reveals
+
+Up to `MAX_EARLY_DERIVED_SECRET_REVEALS`, [`EarlyDerivedSecretReveal`](./custody-game.md#earlyderivedsecretreveal) objects can be included in the `block`. The early derived secret reveals must satisfy the verification conditions found in [early derived secret reveal processing](./custody-game.md#custody-key-reveals). The validator receives a small "whistleblower" reward for each early derived secrete reveal included.
+
+##### Shard transitions
+
+Exactly `MAX_SHARDS` [`ShardTransition`](./beacon-chain#shardtransition) objects are included in the block. Default each to an empty `ShardTransition()`. Then for each committee assigned to the slot with an associated `committee_index` and `shard`, set `shard_transitions[shard] = full_transitions[winning_root]` if the committee had enough weight to form a crosslink this slot.
+
+Specifically:
+* Call `shards, winning_roots = get_successful_shard_transitions(state, block.slot, attestations)`
+* Let `full_transitions` be a dictionary mapping from the `shard_transition_root`s found in `attestations` to the corresponding full `ShardTransition`
+* Then for each `shard` and `winning_root` in `zip(shards, winning_roots)` set `shard_transitions[shard] = full_transitions[winning_root]`
+
+```python
+def get_successful_shard_transitions(state: BeaconState,
+                                     slot: Slot,
+                                     attestations: Attestation) -> Tuple[Sequence[Shard], Sequence[Root]]:
+    shards = []
+    winning_roots = []
+    online_indices = get_online_validator_indices(state)
+    committee_count = get_committee_count_at_slot(state, slot)
+    for committee_index in map(CommitteeIndex, range(committee_count)):
+        shard = compute_shard_from_committee_index(state, committee_index, slot)
+        # All attestations in the block for this committee/shard and current slot
+        shard_attestations = [
+            attestation for attestation in attestations
+            if attestation.data.index == committee_index and attestation.data.slot == slot
+        ]
+        committee = get_beacon_committee(state, state.slot, committee_index)
+
+        # Loop over all shard transition roots, looking for a winning root
+        shard_transition_roots = set([a.data.shard_transition_root for a in shard_attestations])
+        for shard_transition_root in sorted(shard_transition_roots):
+            transition_attestations = [
```

Comment:
From grouping `transition_attestations`, to computing `winning_root`. This can be refactored into its own helper function and reuse in `process_crosslink_for_shard`.  It'll reduce disparity and make testing easier down the road